### PR TITLE
21770 rework sip logging extension

### LINKED ISSUE #21770: Rework the Sip Log Extension
Follow PR https://github.com/OpenLiberty/open-liberty/pull/21465,  the SIP and rtcomm_connector FATs started to encounter consistent failures.  This indicates the [SipLogExtension](https://github.com/OpenLiberty/open-liberty/blob/2fd4a880754c37a988c5ed9ac4f1ea5988e465d6/dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java) needs to be reworked.  Additionally, numerous  sip sessions are created. 

Stack traces of errors: 

```
Stack Dump = java.lang.IllegalStateException: Invalid Session: Id = wlp_53 Info = com.ibm.ws.sip.container.servlets.WASXSipApplicationSessionImpl@b71a52cc
	at com.ibm.ws.sip.container.servlets.SipApplicationSessionImpl.checkIsSessionValid(SipApplicationSessionImpl.java:1302)
	at com.ibm.ws.sip.container.servlets.SipApplicationSessionImpl.getSessions(SipApplicationSessionImpl.java:700)
	at com.ibm.ws.sip.container.servlets.WASXSipApplicationSessionImpl.getSessions(WASXSipApplicationSessionImpl.java:134)
	at com.ibm.ws.sip.container.util.SipLogExtension.getSecondSessionId(SipLogExtension.java:262)
	at com.ibm.ws.sip.container.util.SipLogExtension.access$400(SipLogExtension.java:35)
	at com.ibm.ws.sip.container.util.SipLogExtension$5.getValue(SipLogExtension.java:106)
	at com.ibm.websphere.logging.hpel.LogRecordContext.getExtensions(LogRecordContext.java:308)
	at com.ibm.ws.logging.internal.WsLogRecord.createWsLogRecord(WsLogRecord.java:531)
	at com.ibm.ws.logging.internal.impl.BaseTraceService.exit(BaseTraceService.java:887)
	at com.ibm.websphere.ras.Tr.exit(Tr.java:716)
	at com.ibm.sip.util.log.LogMgr.traceExit(LogMgr.java:200)
	at com.ibm.ws.sip.container.was.ContextBasedQueue.recordThreadID(ContextBasedQueue.java:373)
	at com.ibm.ws.sip.container.was.SignalingEndOfTask.run(SignalingEndOfTask.java:72)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:245)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:866)
```

```
 Exception in thread "Default Executor-thread-33" 
 java.lang.NullPointerException
 	at com.ibm.ws.sip.container.tu.TransactionUserBase.setHandler(TransactionUserBase.java:901)
 	at com.ibm.ws.sip.container.tu.TransactionUserWrapper.setHandler(TransactionUserWrapper.java:1198)
 	at com.ibm.ws.sip.container.servlets.SipSessionImplementation.setHandler(SipSessionImplementation.java:845)
 	at com.ibm.ws.sip.container.servlets.SipSessionImplementation.init(SipSessionImplementation.java:217)
 	at com.ibm.ws.sip.container.servlets.SipSessionImplementation.<init>(SipSessionImplementation.java:157)
 	at com.ibm.ws.sip.container.servlets.SIPSessionFactory.createSIPSession(SIPSessionFactory.java:44)
 	at com.ibm.ws.sip.container.tu.TransactionUserBase.getSipSession(TransactionUserBase.java:937)
 	at com.ibm.ws.sip.container.tu.TransactionUserBase.getAllSipSessions(TransactionUserBase.java:615)
 	at com.ibm.ws.sip.container.tu.TransactionUserWrapper.getAllSipSessions(TransactionUserWrapper.java:1425)
 	at com.ibm.ws.sip.container.servlets.SipApplicationSessionImpl.getAllSIPSessions(SipApplicationSessionImpl.java:681)
 	at com.ibm.ws.sip.container.servlets.SipApplicationSessionImpl.getAllSIPSessions(SipApplicationSessionImpl.java:652)
 	at com.ibm.ws.sip.container.servlets.SipApplicationSessionImpl.getSessions(SipApplicationSessionImpl.java:708)
 	at com.ibm.ws.sip.container.servlets.WASXSipApplicationSessionImpl.getSessions(WASXSipApplicationSessionImpl.java:134)
 	at com.ibm.ws.sip.container.util.SipLogExtension.getSecondSessionId(SipLogExtension.java:262)
 	at com.ibm.ws.sip.container.util.SipLogExtension.access$400(SipLogExtension.java:35)
 	at com.ibm.ws.sip.container.util.SipLogExtension$5.getValue(SipLogExtension.java:106)
 	at com.ibm.websphere.logging.hpel.LogRecordContext.getExtensions(LogRecordContext.java:308)
 	at com.ibm.ws.logging.internal.WsLogRecord.createWsLogRecord(WsLogRecord.java:531)
 	at com.ibm.ws.logging.internal.impl.BaseTraceService.error(BaseTraceService.java:806)
 	at com.ibm.websphere.ras.Tr.error(Tr.java:629)
 	at com.ibm.sip.util.log.LogMgr.error(LogMgr.java:297)
 	at com.ibm.ws.sip.container.was.SignalingEndOfTask.run(SignalingEndOfTask.java:93)
 	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:245)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
 	at java.base/java.lang.Thread.run(Thread.java:866)
```

### PR DESCRIPTION
**Pull request by TiborBeres**
When HPEL logging is enabled, the SipLogExtension class is calling getSessions("SIP") method, which results is the sipcontainer creating new application sessions when none exist. This behavior is wrong, logging activity should not result in the creation of new sipp application sessions. 

fixes  #21770

### CODE REVIEW

**Comment by TiborBeres** [comment]
#build

**Comment by TiborBeres** [comment]
#build

File: dev/com.ibm.websphere.javaee.servlet.sip.1.1/src/javax/servlet/sip/SipApplicationSession.java [hunk 1]
```diff
@@ -252,6 +252,27 @@ Object getSession(java.lang.String id,
...
+     * @param protocol a string identifying the protocol name, e.g. "SIP"
+     * @param boolean representing if SIP sessions should be created if none exist
+     * @return <code>Iterator</code> over protocol sessions of the
+     *      specified protocol  
+     * @throws IllegalStateException - if this application session is not valid
+     * @throws NullPointerException - if the protocol is null 
+     * @throws IllegalArgumentException - if the protocol is not understood by container.
+     * 
+     */
+    Iterator getSessions(String string, boolean b) throws IllegalStateException, NullPointerException, IllegalArgumentException; ;
```
Comment (volosied) [inline dev/com.ibm.websphere.javaee.servlet.sip.1.1/src/javax/servlet/sip/SipApplicationSession.java]: Just for readability, could you update the params names to `String protocol, boolean create`, same as in the implementation.

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java [hunk 2]
```diff
@@ -1800,4 +1806,36 @@ public void setSessionKeyBaseTargeting(String skbt){
 		}
 		m_sessionKeyBaseKey = skbt;
 	}
+	
+	public Iterator getSessions(String protocol, boolean b) {
```
Comment (volosied) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java]: Ah, you use `protocol` here. Could you update  `b` to `create`?

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java [hunk 3]
```diff
@@ -1800,4 +1806,36 @@ public void setSessionKeyBaseTargeting(String skbt){
...
+		}
+		
+		else if (protocol.equalsIgnoreCase("HTTP")){
+			return Collections.EMPTY_MAP.keySet().iterator();
+		}
+			
+		else { 
+
+			SipAppDesc sipAppDesc = getAppDescriptor();
+			if(sipAppDesc != null && !sipAppDesc.isJSR289Application()){
```
Comment (volosied) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java:1829]: I'm not familiar with this. What does this mean?  I guess what I mean under what condition would you not throw the IllegalArgumentException?

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/tu/TransactionUserBase.java [hunk 4]
```diff
@@ -611,19 +611,17 @@ private static String createApplicationId( String srvrid){
...
 		if(_sipSession != null){
 			sessions.add(_sipSession);	
 		}
 		return sessions;
 	}
 
-	/**
-	 * Invalidate the Transaction User and Remove it from the TransactionUserTable
-	 *
-	 */
```
Comment (volosied) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/tu/TransactionUserBase.java]: Unrelated, but was this comment removed on purpose?

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java [hunk 5]
```diff
@@ -141,12 +140,15 @@ public static void destroy() {
...
-    	
+    	try {
     	SipApplicationSession sas = ThreadLocalStorage.getApplicationSession();
     	if (sas != null) {
     		return sas.getId(); 
     	}
     	
+    	} catch (NullPointerException e) {
+    		
+    	}
```
Comment (volosied) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java:151]: What are the NPE catches for?

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java [hunk 6]
```diff
@@ -669,23 +669,28 @@ protected List<IBMSipSession> getAllSIPSessions(boolean create)
...
+						result.addAll(tu.getAllSipSessions(create));
+						}
+			
+						if (c_logger.isTraceDebugEnabled()) {
+							c_logger.traceDebug(this, "getAllSIPSessions", "found SIP sessions. count:"+result.size());
+						}
+					}
 			return result;
+			}
+
```
Comment (volosied) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java]: Just a general comment: could you fix the indentation / formatting here and a few other places as some of it a bit hard to follow. Which brackets match.

**Review by volosied — COMMENTED** [review]
Looks good over all, The change make sense to address  the NPE / session creating in the logging we had before.

Just a few comments and questions.

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java (hunk 5 above)
Comment (TiborBeres) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java:151]: the logger code was throwing NPEs before I wrapped that section of code into try / catch

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/tu/TransactionUserBase.java (hunk 4 above)
Comment (TiborBeres) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/tu/TransactionUserBase.java]: comment was not removed on purpose, I guess it got deleted somewhere with the several versions of code I tried along the way. I will re-add the comment

File: dev/com.ibm.websphere.javaee.servlet.sip.1.1/src/javax/servlet/sip/SipApplicationSession.java (hunk 1 above)
Comment (TiborBeres) [reply dev/com.ibm.websphere.javaee.servlet.sip.1.1/src/javax/servlet/sip/SipApplicationSession.java]: parameter names changed both in SipApplicationSession and SipApplicationSessionImpl

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java (hunk 2 above)
Comment (TiborBeres) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java]: changed b to protocol

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java (hunk 3 above)
Comment (TiborBeres) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java:1829]: sipcontainer only knows about SIP and HTTP protocols. We are not restricting what string will be passed into this getSessions(String, boolean) method, but we are only handling SIP and HTTP. If anything else is passed in this method, the IllegalArgumentException will be throws. This is the same in the previously existing getSessions(String) method.

**Comment by TiborBeres** [comment]
#build

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java (hunk 3 above)
Comment (volosied) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java:1829]: I guess what's confusing me here is that you can pass in a bad protocol, but it could still return a iterator here because of `sipAppDesc != null && !sipAppDesc.isJSR289Application()` .

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/tu/TransactionUserBase.java (hunk 4 above)
Comment (volosied) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/tu/TransactionUserBase.java]: This comment wasn't added back in?

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java [hunk 7]
```diff
@@ -188,6 +190,9 @@ private static String getCallId() {
 			}
 		}
 		
+    	} catch (NullPointerException e ) {
+    		
+    	}
```
Comment (volosied) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java:198]: Can you add a comment to explain the npe catches one from the logging?  Here an the other locations?

**Review by volosied — APPROVED** [review]

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java (hunk 5 above)
Comment (volosied) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java:151]: Should have replied here instead about adding comments as it's just not clear by the code alone.

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java (hunk 3 above)
Comment (TiborBeres) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java:1829]: it is the same logic we had in place all along. I don't think we should modify it without a good reason.

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java (hunk 7 above)
Comment (TiborBeres) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/util/SipLogExtension.java:198]: Comments added

**Comment by TiborBeres** [comment]
#libby

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java (hunk 3 above)
Comment (volosied) [reply dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/servlets/SipApplicationSessionImpl.java:1829]: I complete missed that. I see now! thanks
